src/components/datasource.vue

```diff
@@ -0,0 +1,452 @@
+                </div>
+              </div> <!-- uk-margin -->
+            </div> <!-- column 1 -->
+            <div class="uk-width-1-2">
+              <div class="uk-margin">
+                <label class="uk-form-label" for="data-fields">Available Fields</label>
+                <div class="uk-form-controls">
+                  <div class="uk-overflow-auto uk-height-medium" style="border:1px solid #e5e5e5">
+                    <table class="uk-table uk-table-small uk-table-divider uk-table-hover">
+                      <tbody>
+                        <tr v-for="field in fields">
+                          <td class="uk-table-link">
+                            <a class="uk-link-reset" href="" @click.stop.prevent @dblclick='onInsertField(field)'>
+                              {{ field.label }}
+                            </a>
+                          </td>
+                        </tr>
+                      </tbody>
+                    </table>
+                  </div> <!-- overflow -->
+                </div>
+              </div>
+            </div> <!-- column 2 -->
+          </div>
+        </li> <!-- popover tab container -->
+        <li>
+          <div class="uk-panel uk-width-1-1">
+            <div class="leaflet-pane leaflet-popup-pane">
+              <div class="leaflet-popup">
+                <div class="leaflet-popup-content-wrapper">
+                  <div class="leaflet-popup-content" v-html="samplePopover" id="samplePopover">
+                  </div>
+                </div>
+                <div class="leaflet-popup-tip-container">
+                  <div class="leaflet-popup-tip"></div>
+                </div>
+                <a class="leaflet-popup-close-button" @click.stop.prevent href="#close">×</a>
+              </div>
+            </div>
+            <div class="uk-panel uk-text-center">
+              <img :src="selectedIcon" width="32" height="32">
+            </div>
+          </div>
+        </li> <!-- preview tab container -->
+      </ul>
+
+      <div class="uk-panel uk-margin">
+        <div class="uk-float-right">
+          <button
+            type="button"
+            class="uk-button uk-button-default"
+            @click="onCancel">Cancel</button>
+          <button
+            type="button"
+            class="uk-button uk-button-primary"
+            @click="onSave">Save</button>
+        </div>
+      </div>
+    </form>
+  </div>
+</template>
+
+<script>
+import DataConnector from './dataconnector.vue'
+import { validationMixin } from 'vuelidate'
+import { required } from 'vuelidate/lib/validators'
+import nanoid from 'nanoid'
+import store from './store'
+import templayed from 'templayed'
+import { ArcGISLayer } from '../lib/arcgis'
+import { debounce } from 'lodash'
+
+export default {
+  name: 'data-source',
+  mixins: [validationMixin],
+  components: {
+    DataConnector,
+  },
+  props: {
+    uid: {
```

Comment:
I think there's a missed opportunity here to take advantage of the fact that you’re backed by a Vuex store.

I would recommend taking in a data source object from your props and then using `_.deepClone` and setting that as a `data` property, which will re-add the reactivity. Then, you can manipulate the `data` version of the data source without affecting the original one from the store, and when you need to save you just ship your `data` to the store with  an `updateDataSource` action.

There's a lot of destructuring the data source into attributes throughout this code, which makes it fragile when it comes time to add or change the properties of a data source.

Reply:
Yea, it shows that I added that so late in the process. I was trying to be lightweight starting out, but finally caved in.

Let's put this on the refactor list.